src/TreeView/index.tsx

```diff
@@ -612,18 +619,11 @@ const TreeView = React.forwardRef<HTMLUListElement, ITreeViewProps>(
       </ul>
     );
   }
-  // TS can't differentiate between JSX and generics in TSX unless we extend something
-  // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-constraint
-) as (<M extends unknown = IMetadata>(
-  props: PropsWithoutRef<ITreeViewProps<M>>,
-  ref: RefAttributes<HTMLUListElement>
-) => ReactElement | null) & {
-  propTypes: WeakValidationMap<PropsWithoutRef<ITreeViewProps>>;
-};
+) as TreeViewComponent;
```

Comment:
Please don’t use type conversions. The function should be written such that the type is determined automatically without the use of as.

If you’re having trouble due to the forward ref, I recall that there is a way to make a type-aware forward ref, but I don’t have the doc for that off the top of my head.

Reply:
This is the correct way to make a forward ref that accepts a generic. All I've done is abstracted it into a single type so that we can also define the type of `propTypes`. The other way of doing this is to augment the `forwardRef` function directly on the `react` module, which in my opinion is worse as it will modify the type of `forward` ref globally

https://stackoverflow.com/questions/58469229/react-with-typescript-generics-while-using-react-forwardref

Reply:
Without this, you can't have `TreeView` accept a generic which means that `TreeView` can't infer the type of `metadata`. The only way to define the type of `metadata` without this would be to explicitly define it on the `INode` or in `flattenTree`.

Reply:
The top answer in [the link you provided ](https://stackoverflow.com/questions/58469229/react-with-typescript-generics-while-using-react-forwardref)has an alternative that is what I have used successfully in the past. It’s option 2

```ts
const FRefOutputComp2 = React.forwardRef(FRefInputComp)
// ↳ T is instantiated with base constraint `Option<unknown>` from FRefInputComp

export const Wrapper = <T extends Option>({myRef, ...rest}: Props<T> & 
  {myRef: React.Ref<HTMLDivElement>}) => <FRefOutputComp2 {...rest} ref={myRef} />

const Usage2 = () => <Wrapper options={options} myRef={myRef} />
```

There’s a tradeoff of the type conversion that people don’t think about which is that the the thing being converted can change and TS isn’t always smart enough to notice that the conversion isn’t accurate anymore which can cause runtime errors. So that’s why I avoid the as keyword at all times.

I hope that option 2 will keep the generic. It’s been a while since I had to do this myself.

But anyway, take your time since we’re not going to merge this until a member of @mellis481’s team tries this out in our test bed library to ensure that the changes work.

Reply:
The problem with option 2 is that it is a breaking change. `TreeView` would no longer support the `ref` prop as you are forced to replace it with a renamed prop. Something like `innerRef` is commonly used. As long as you are fine with the breaking change, then we can use option 2.

My general goal when submitting PRs is to avoid breaking changes where possible.

Reply:
> But anyway, take your time since we’re not going to merge this until a member of @mellis481’s team tries this out in our test bed library to ensure that the changes work.

Test bed library? You don't rely on the the library's own unit tests?

Reply:
@dgreene1 Any thoughts on my previous 2 comments?

Reply:
I stand by my answer above and there’s a way to do it without a breaking change and without type casting.